src/jvm/main/org/jetbrains/kotlinx/lincheck/nvm/crash.kt

```diff
@@ -0,0 +1,200 @@
+/*-
+ * #%L
+ * Lincheck
+ * %%
+ * Copyright (C) 2019 - 2020 JetBrains s.r.o.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.nvm
+
+import kotlinx.atomicfu.atomic
+import org.jetbrains.kotlinx.lincheck.runner.TestThreadExecution
+import java.util.*
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * This exception is used to emulate system crash.
+ * Must be ignored by user code, namely 'catch (e: Throwable)' constructions should pass this exception.
+ */
+abstract class CrashError(enableStackTrace: Boolean) : Throwable(null, null, false, enableStackTrace) {
+    var actorIndex: Int = -1
+    abstract val crashStackTrace: Array<StackTraceElement>
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CrashError) return false
+
+        if (actorIndex != other.actorIndex) return false
+        if (!crashStackTrace.contentEquals(other.crashStackTrace)) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = actorIndex
+        result = 31 * result + crashStackTrace.contentHashCode()
+        return result
+    }
+}
+
+class CrashErrorImpl : CrashError(true) {
+    override val crashStackTrace: Array<StackTraceElement> get() = stackTrace
+}
+
+/** Proxy provided to minimize [fillInStackTrace] calls as it influence performance a lot. */
+class CrashErrorProxy(private val ste: StackTraceElement?) : CrashError(false) {
+    override val crashStackTrace get() = if (ste === null) emptyArray() else arrayOf(ste)
+}
+
+private data class SystemContext(
+    val waitingThreads: Int,
+    val threads: Int,
+    val free: AtomicBoolean = AtomicBoolean(true)
+)
+
+object Crash {
+    private val systemCrashOccurred = AtomicBoolean(false)
+    private val context = atomic(SystemContext(0, 0))
+    private var awaitSystemCrashBeforeThrow = true
+    internal val threads get() = context.value.threads
+
+    @Volatile
+    private var execution: TestThreadExecution? = null
```

Comment:
Yes, as we have to call `incClock` manually (and before opening the barrier!) in case of durable linearizability to be sure that all clocks are updated at the next `readClocks`